Add backwards compatible database manager which allows per plugin config

minnsoe opened the pull request:
Resolves https://github.com/backstage/backstage/issues/2894 by introducing a backwards compatible database manager.

## Changes

A new backwards compatible database manager, `PluginConnectionDatabaseManager`, allows the end-user to set global and per plugin database configuration. The PR also introduces an early iteration of a database connector interface and provides helper functions for each of the database connectors to meet the new interface.

## Example
Given the `backend.database` configuration below:
```yaml
backend:
  database:
    client: pg
    prefix: test_prefix_
    connection:
      host: localhost
      user: postgres
      password: backstage
    plugin:
      code-coverage:
        connection:
          database: pg_code_coverage_set_by_user
```

The implementation uses/creates the following databases:

<img width="860" alt="image" src="https://user-images.githubusercontent.com/1093775/118498079-8ea45100-b71d-11eb-863f-fd83d54412a0.png">

## Usage

Existing backstage installations can be migrated by swapping out the database
manager under `packages/backend/src/index.ts` as shown below:

```diff
import {
-  SingleConnectionDatabaseManager,
+  PluginConnectionDatabaseManager,
} from '@backstage/backend-common';

// ...

function makeCreateEnv(config: Config) {
  // ...
-  const databaseManager = SingleConnectionDatabaseManager.fromConfig(config);
+  const databaseManager = PluginConnectionDatabaseManager.fromConfig(config);
  // ...
}
```

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

minnsoe commented:
Just noticed a prettier warning on the `md` file from the workflow. I've fixed that and rebased back onto master.

freben reviewed (commented):
Hey, this is really well made and I appreciate the contribution and the functionality that it aims to solve for. I also made a first quick scan and left a few comments here and there.

But what I'd mostly like to open up for discussion, is the approach of having a second class, and to drive it via config. Bear with me!

There's an ongoing balancing act of what to drive in config, and what to drive in code. There are some cases where config is appropriate (sane default behaviors and very-often-tweaked basic things) or required (secrets). However, our general gut feeling lately is that the threshold for when it's more convenient to abandon config and switch to code-driven tweaking, is actually pretty low.

So what would that mean in terms of this PR?

I started thinking that maybe it could make sense to leave the config unchanged, and keep the one `SingleConnectionDatabaseManager`, but enable users to do the following:

```ts
const databaseManager = SingleConnectionDatabaseManager
  .fromConfig(config)
  .withPrefix('my_company_')
  .withPluginOverride('catalog', {
    connection: { database: 'totally_custom' }
  })
```

I think this is as least as expressive, gives the benefit of typescript editor support aiding you in writing all of the right things, and would probably amount to a lot less code and special-casing.

The remaining problem would be if you _actually_ needed config to drive it - which essentially would only be the case if you used secrets. And those, you can always put in custom config keys and load them before calling `withPluginOverride`.

This is not really a work order, but a discussion opener. I'll ping @backstage/maintainers here, because I think it's worth getting more input on.

minnsoe commented:
Hey @freben, thank you for reviewing this. I'm glad you've brought up those points because they were also on my mind when putting this forward. When implementing this, I took care to ensure the interface is a superset of the existing one. In theory, we can effectively alias `SingleConnectionDatabaseManager` to `PluginConnectionDatabaseManager` without the end-user noticing the difference.

On the balancing act of what to drive in config, it was precisely the use case you mentioned which drove this change. We needed to change the connection in addition to the database names due to our declarative infra which includes credentials. A fluent API structure makes a lot of sense, in fact I'm quite keen to adopt the `withPrefix` changes.

On the config driven `backend.database.plugin.<pluginId>.connection` override section, I see some benefit in having a designated block/pattern to put the plugin database config (to get secrets populated via environment variables) but perhaps not actually driving the behaviour as it is somewhat implicit.

What do you think about a hybrid? I'm thinking of the following:

- Rename `PluginConnectionDatabaseManager` to `DatabaseManager`.
- Deprecate and alias `SingleConnectionDatabaseManager` to `DatabaseManager` and type cast to limit TS info.
- Add fluent API method `.withPrefix` to `DatabaseManager` and drive that as code and completely drop from config.
- Add fluent API method `.withPluginOverride` to `DatabaseManager` with the same shape as you've mentioned above.
- Add fluent API method `.withPluginConfig` to `DatabaseManager` which takes in a `pluginId` and is responsible for applying plugin config from `backend.database.plugin.<pluginId>`. Under the hood this calls `.withPluginOverride` and reads the config for you.

With the last method, I believe this would allow developers to set basic plugin database config as below instead of always needing to read it in themselves for simple use cases:

```typescript
import { DatabaseManager } from '@backstage/backend-common';
import { Config } from '@backstage/config';

function makeCreateEnv(config: Config) {
  // ...
  const databaseManager = DatabaseManager
    .fromConfig(config)
    .withPluginConfig('catalog');
```
as a convenience method for doing the following:

```typescript
import { DatabaseManager } from '@backstage/backend-common';
import { Config } from '@backstage/config';

function makeCreateEnv(config: Config) {
  // ...
  const databaseManager = DatabaseManager
    .fromConfig(config)
    .withPluginOverride('catalog', {
      connection: {
        database: config.getString('some.config.for.database'),
        password: config.getString('some.config.for.password'),
      },
    });
```

Naturally, the set of fluent APIs above gives back control to the developer and we can also provide other (slightly more opinionated) methods where the config needs to sit under `backend.database.plugin.*`. On the topic of merging, the order will now be in the developer's control. e.g. if you call `.withPluginOverride` first and `.withPluginConfig` second - the plugin config will be applied last and override earlier values. The developer can choose which order and for which plugins via code.

What do you think? Would you be opposed with going down the `.withPrefix`,`.withPluginOverride` and `.withPluginConfig` path?

Rugvip commented:
@minnsoe We're leaning back towards treating the database setup as a completely config-driven thing at this point to be honest. Meaning we don't expect people to provide a different database manager implementations except for perhaps in some extreme cases. The biggest reason for this is how tightly all parts of the database configuration is tied to things that require static configuration: endpoints, secrets, etc. I think DX ends up being pretty bad when we have a config schema in place that is only enabled when it's also mirrored with the specific code that also reads that config.

With that said I suggest the following: rename `SingleConnectionDatabaseManager` to `PluginDatabaseManager` and implement the full prefix and config override functionality there. Add an alias for backwards compatibility and mark the aliased type as deprecated, pointing to the new name instead. Skip the fluid interface, keep `fromConfig` as the single method.

Suggesting to drop `Connection` from `PluginConnectionDatabaseManager` because it feels a bit redundant, previously there was emphasis on a single connection, but that's now gone.

minnsoe commented:
@Rugvip Okay, great. I'll make the changes to deprecate `SingleConnectionDatabaseManager`, alias and present a single config driven `PluginDatabaseManager`.

Rugvip reviewed (commented):
Thanks! Looks good, couple of nits and I think it's best to move out the MySQL addition

minnsoe commented on docs/tutorials/configuring-plugin-databases.md:
```diff
@@ -0,0 +1,123 @@
+tools such as Terraform or AWS CloudFormation where the name of each database is
+defined, created and assigned explicitly. You may also need to use different
+credentials for each database or use a set of credentials without the
+permissions needed to create databases.
```
Thanks for the review @freben. I've made sure to call out the control of the database name part explicitly in the updated version of the docs for clarity:

https://github.com/backstage/backstage/pull/5700/files#diff-18d29dfb4a92d93a741c679b9c95fe814f6baf26074281d48688210a1e538276R21-R25

minnsoe commented on docs/tutorials/configuring-plugin-databases.md:
```diff
@@ -0,0 +1,123 @@
+- **`backend.database.plugin.<pluginId>`:** is used to define a `client` and
+  `connection` block for the plugin matching the `pluginId`, e.g. `catalog` is
+  the `pluginId` for the catalog plugin and any configuration defined under that
+  block is specific to that plugin.
```
Good point about documenting the behaviour, resolved in latest diff: 

https://github.com/backstage/backstage/pull/5700/files#diff-18d29dfb4a92d93a741c679b9c95fe814f6baf26074281d48688210a1e538276R77-R82

minnsoe commented on packages/backend-common/src/database/DatabaseManager.ts:
```diff
@@ -0,0 +1,160 @@
+      // attempt to lookup sqlite3 database file name
+      pluginConfig.getOptionalString('connection.filename') ??
+      // if root is sqlite - attempt to use top level connection, fallback to :memory:
+      (isSqlite ? rootSqliteName : null) ??
```
Addressed, I've split up the plugin block so that there are clear logical branches to follow.

https://github.com/backstage/backstage/pull/5700/commits/134f98d0fa12cb7e5ab6ca06ebe81b1128e19b44

minnsoe commented on packages/backend-common/src/database/connection.ts:
```diff
@@ -14,14 +14,30 @@
+import { pgConnector } from './postgres';
+import { sqlite3Connector } from './sqlite3';
+
+type DatabaseClient = 'pg' | 'sqlite3' | 'mysql' | 'mysql2' | string;
```
Agreed, probably a good idea to work on additional improvements or behaviour with a seperate PR. I presume this is regarding `mysql` vs `mysql2`?

minnsoe commented on packages/backend-common/src/database/connector.ts:
```diff
@@ -0,0 +1,30 @@
+import { Config } from '@backstage/config';
+import { Knex } from 'knex';
+
+export interface DatabaseConnector {
```
Resolved, moved the interface into `types.ts` and I've also reorganized the connectors into their own module.

minnsoe commented on packages/backend-common/src/database/mysql.ts:
```diff
@@ -159,3 +160,23 @@ export async function ensureMysqlDatabaseExists(
   }
 }
+
+export function createMysqlNameOverride(name: string): Partial<Knex.Config> {
```
Resolved, `connectors/defaultNameOverride.ts` handles this now and duplication has been deleted.

minnsoe commented:
@Rugvip, the logic in the database manager itself has been refactored and documented for more clarity. I've also added docstrings to the rest of the interfaces, methods and functions.

With regards to the `mysql` connector, I've made the code paths inaccessible but I've preserved the implementation because I believe it was introduced by @freben as prep work for future support. The inverse of change set, https://github.com/backstage/backstage/pull/5700/commits/03a7b0713ae5fec71a6d60aee11726bb5a1b16d1, should hook it back up once ready.

Thanks for the review!

freben commented:
Oh no - a conflict crept in. I'll take a look through the PR as a whole now again too.

minnsoe commented:
@freben, addressed the conflict and rebased.

I've also fixed another test elsewhere by prefixing the integer test plugin names which were being generated by `backend-test-utils` with `db`. It was breaking because `Config` does not permit config keys which start with a number: https://github.com/backstage/backstage/blob/c8a70893a260cf609605093d9a629380cd74aeba/packages/config/src/reader.ts#L22

minnsoe commented:
Looking at the last CI run, I've dropped the commit which drops mysql from being hooked up and rebased again. The basis for this is that it looks like we have code and tests in `backend-test-utils` which are exercising the mysql code paths

https://github.com/backstage/backstage/pull/5700/checks?check_run_id=2777723825#step:20:460

minnsoe commented:
@freben, fixed three references in `backend-test-utils`: `backstage_plugin_0` => `backstage_plugin_db0`.

> FAIL src/database/TestDatabases.test.ts (71.624 s)
>  ● TestDatabases › obeys a provided connection string for postgres 13
>
>    expect(received).resolves.toEqual()
>
>    Received promise rejected instead of resolved
>    Rejected to value: [error: database "backstage_plugin_0" does not exist]

freben commented:
It's a bit annoying in this case that we aren't running the builds unless I press a button. Makes the feedback loop slower. After we get this nice contribution in, the next one will be smoother!

freben commented on .changeset/five-donkeys-brake.md:
```diff
@@ -0,0 +1,50 @@
+---
+'@backstage/backend-common': minor
```
need to add backend-test-utils here now, don't we

freben reviewed (commented):
Alright, this is looking great! One single small thing left - you missed adding a changeset line which is important now that the updated backend-test-utils is in need of shipping at the same time. Then let's :shipit:

minnsoe commented:
> It's a bit annoying in this case that we aren't running the builds unless I press a button. Makes the feedback loop slower. After we get this nice contribution in, the next one will be smoother!

Definitely. It's a bit of a shame that we need to treat these sort of things as hostile. I noticed another PR which looked like spam which could have changed the pipeline to siphon off compute for other purposes.

minnsoe commented:
Updated the changeset and rebased. Looking forward to shipping it! 🚢

minnsoe commented on .changeset/five-donkeys-brake.md:
```diff
@@ -0,0 +1,51 @@
+---
+'@backstage/backend-common': minor
+'@backstage/create-app': minor
+'@backstage/backend-test-utils': minor
```
Fixed.

minnsoe commented on packages/backend-common/api-report.md:
```diff
@@ -100,6 +104,20 @@ export function createServiceBuilder(_module: NodeModule): ServiceBuilderImpl;
 export function createStatusCheckRouter(options: StatusCheckRouterOptions): Promise<express.Router>;
 
+// @public
+export interface DatabaseConnector {
```
@Rugvip, I agree and it makes it easier to tweak in the future too. I've dropped the export of the `DatabaseConnector` interface externally.

About the other helper functions, these were exported with a glob and additional functions were implemented in this PR: https://github.com/backstage/backstage/pull/5700/files#diff-9c7833c9467a96ce53bf48256288be421fcc2c5d57d065b1bb7a8f62833b8a11L17 Just to confirm that we want to close up this (undocumented) API including the others inside?

Some folks might be using the undocumented API, but this MR should provide the functionality so it shouldn't be such a big issue: https://github.com/backstage/backstage/issues/2894#issuecomment-801503125

Rugvip replied:
Let's keep the existing noes in place but not add any new ones. We'll want to deprecate those too later though

minnsoe replied:
Done, I've pulled out the existing exports from there explicitly, added a note and regenerated the api-report.

minnsoe commented:
Thanks for keeping the PR fresh. Linear history would have been nice but it would be a stream of linear in a sea of merge commits! 🌊

Rugvip commented:
@minnsoe yep linear is 👌, bit GitHub makes that difficult at times 😅

Rugvip reviewed (approved):
Awesome, let's :shipit:! 😁
